File: src/NeuralDamage.Domain/Reaction.cs
namespace NeuralDamage.Domain;

public class Reaction : BaseEntity
{
    public required Guid MessageId { get; init; }
    public Guid? UserId { get; init; }
    public Guid? BotId { get; init; }
    public required string Emoji { get; set; }

    public Message Message { get; set; } = null!;
    public User? User { get; set; }
    public Bot? Bot { get; set; }
}

File: backend/app/models/reaction.py
from datetime import datetime
from uuid import uuid4

from sqlalchemy import String, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Reaction(Base):
    __tablename__ = "reactions"
    __table_args__ = (
        UniqueConstraint("message_id", "user_id", "emoji", name="uq_reaction_user"),
        UniqueConstraint("message_id", "bot_id", "emoji", name="uq_reaction_bot"),
    )

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    message_id: Mapped[str] = mapped_column(String(36), ForeignKey("messages.id", ondelete="CASCADE"), index=True)
    user_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    bot_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("bots.id"), nullable=True)
    emoji: Mapped[str] = mapped_column(String(32), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    message = relationship("Message", back_populates="reactions")
    user = relationship("User")
    bot = relationship("Bot")
